lib/Controller/Report.php

```diff
@@ -0,0 +1,520 @@
+
+        $this->store = $store;
+        $this->timeSeriesStore = $timeSeriesStore;
+        $this->reportService = $reportService;
+        $this->reportScheduleFactory = $reportScheduleFactory;
+        $this->savedReportFactory = $savedReportFactory;
+        $this->mediaFactory = $mediaFactory;
+        $this->layoutFactory = $layoutFactory;
+    }
+
+    /// //<editor-fold desc="Report Schedules">
+
+    /**
+     * Report Schedule Grid
+     */
+    public function reportScheduleGrid()
+    {
+        $reportSchedules = $this->reportScheduleFactory->query($this->gridRenderSort(), $this->gridRenderFilter([
+            'name' => $this->getSanitizer()->getString('name'),
+        ]));
+
+        /** @var \Xibo\Entity\ReportSchedule $reportSchedule */
+        foreach ($reportSchedules as $reportSchedule) {
+
+            $reportSchedule->includeProperty('buttons');
+
+            $cron = \Cron\CronExpression::factory($reportSchedule->schedule);
+            $nextRunDt = $cron->getNextRunDate(\DateTime::createFromFormat('U', $reportSchedule->lastRunDt))->format('U');
+            $reportSchedule->nextRunDt = $nextRunDt;
+
+            $reportSchedule->reportName = $this->reportService->getReportByName($reportSchedule->reportName)->description;
+
+            switch ($reportSchedule->schedule) {
+
+                case ReportSchedule::$SCHEDULE_DAILY:
+                    $reportSchedule->schedule = 'Run once a day, midnight';
+                    break;
+
+                case ReportSchedule::$SCHEDULE_WEEKLY:
+                    $reportSchedule->schedule = 'Run once a week, midnight on Monday';
+
+                    break;
+
+                case ReportSchedule::$SCHEDULE_MONTHLY:
+                    $reportSchedule->schedule = 'Run once a month, midnight, first of month';
+
+                    break;
+
+                case ReportSchedule::$SCHEDULE_YEARLY:
+                    $reportSchedule->schedule = 'Run once a year, midnight, Jan. 1';
+
+                    break;
+            }
+
+            // Edit
+            $reportSchedule->buttons[] = [
+                'id' => 'reportSchedule_edit_button',
+                'url' => $this->urlFor('reportschedule.edit.form', ['id' => $reportSchedule->reportScheduleId]),
+                'text' => __('Edit')
+            ];
+
+            // Delete
+            $reportSchedule->buttons[] = [
+                'id' => 'reportSchedule_delete_button',
+                'url' => $this->urlFor('reportschedule.delete.form', ['id' => $reportSchedule->reportScheduleId]),
+                'text' => __('Delete')
+            ];
+        }
+
+        $this->getState()->template = 'grid';
+        $this->getState()->recordsTotal = $this->reportScheduleFactory->countLast();
+        $this->getState()->setData($reportSchedules);
+    }
+
+    /**
+     * Report Schedule Add
+     */
+    public function reportScheduleAdd()
+    {
+
+        $this->getLog()->debug('Add Report Schedule');
```

Comment:
Should some of this logic be in the report class? Not all reports will have a type/layout for example?